src/sympc/protocol/falcon/falcon.py

```diff
@@ -63,3 +64,100 @@ def __eq__(self, other: Any) -> bool:
             return False
 
         return True
+
+    @staticmethod
+    def mul_master(x, y, session) -> ReplicatedSharedTensor:
+        """Master method for multiplication.
+
+        Performs Falcon's mul implementation, gets and reshares mul results and distributes shares.
+        This needs to be improved in future, it relies on orchestrator being a trusted third party.
+        Falcon, requires parties to be able to communication between each other.
+
+        Args:
+            x (MPCTensor): Secret
+            y (MPCTensor): Another secret
+            session (Session): Session the tensors belong to
+
+        Returns:
+            shares (ReplicatedSharedTensor): Shares in terms of ReplicatedSharedTensor.
+
+        Raises:
+            ValueError: Raised when number of parties are not three.
+            NotImplementedError: Raised when implementation not present
+
+        """
+        if len(session.parties) != 3:
+            raise ValueError("Falcon requires 3 parties")
+
+        result = None
+
+        if session.protocol.security_type == "semi-honest":
+            result = Falcon.multiplication_protocol(x, y)
+
+        elif session.protocol.security_type == "malicious":
+            raise NotImplementedError(
+                "Multiplication has not been implemented for malicious setting"
+            )
+
+        else:
+            raise NotImplementedError(
+                "Multiplication has not been implemented for "
+                + session.protocol.security_type
+                + " setting"
+            )
+
+        random0 = (
+            session.session_ptrs[0]
+            .przs_generate_random_share(shape=x.shape)
+            .get_shares()
+            .get()[0]
+        )
+        random1 = (
+            session.session_ptrs[1]
+            .przs_generate_random_share(shape=x.shape)
+            .get_shares()
+            .get()[0]
+        )
+        random2 = (
+            session.session_ptrs[2]
+            .przs_generate_random_share(shape=x.shape)
+            .get_shares()
+            .get()[0]
+        )
+
+        # Add random mask to vals
+        shares = [
+            result[0].get() + random0,
+            result[1].get() + random1,
+            result[2].get() + random2,
+        ]
+
+        shares = ReplicatedSharedTensor.distribute_shares(shares, x.session)
+        return shares
+
+    @staticmethod
+    def multiplication_protocol(
+        x: MPCTensor, y: MPCTensor
+    ) -> List[ReplicatedSharedTensor]:
+        """Implementation of Falcon's multiplication with semi-honest security guarantee.
+
+        Args:
+            x (MPCTensor): Secret
+            y (MPCTensor): Another secret
+
+        Returns:
+            shares (ReplicatedSharedTensor): Shares in terms of ReplicatedSharedTensor.
+        """
+        z_shares = []
+
+        for party in range(0, len(x.share_ptrs)):
+
+            z_value = (
```

Comment:
MAYBE : Can we delegate the multiplication functionality to the respective tensor, similar like what we do in `ShareTensor`, as like the parties could do the functionalities, the orchestrator could act only a control center, this would be helpful in P2P setting also.So that we can realize the functionality from tensor level itself.